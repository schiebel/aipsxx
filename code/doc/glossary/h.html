<!DOCTYPE HTML PUBLIC "+//ISBN 82-7640-037::WWW//DTD HTML//EN//2.0" "html.dtd">
<HTML>

<HEAD>

<SCRIPT LANGUAGE = "JavaScript1.2"
src="../javascripts/menu.js"></script>
<SCRIPT LANGUAGE = "JavaScript1.2"
src="../javascripts/aipsmenu.js"></script>
<SCRIPT LANGUAGE = "JavaScript1.2">
<!--
 // Function need to place directory relative to docs tree.
 aMenu = new aipsMenu("..", "#00009C");
 function onLoad() {
    if (document.layers) {
        aMenu.loadMenus();
    }
 }
//-->
</SCRIPT>




<!--leave the BASE comment line in place for local testing-->
<!--BASE HREF="www.cv.nrao.edu/aips++/docs/glossary/h.html"-->
<TITLE>AIPS++ Glossary</TITLE>
</HEAD> 
<body background="../gif/watermark.gif" bgcolor="#ffffff" text="#000000"
link="#cc0f0f" vlink="#00009c">

<table bgcolor="#00009C" cellpadding=4 cellspacing=0 border=0 width=600
height=25>
<tr><td>
<a href="../gettingstarted.html"
onMouseOver="aMenu.showAipsMenu('gsMenu')">
<font size=2 face=helvetica  color="#ffffff">
<b>Getting Started</b></a></td>
<td>
<a href="../user/documentation.html"
onMouseOver="aMenu.showAipsMenu('userMenu')">
<font size=2 face=helvetica color="#ffdd00">
<b>Documentation</b></a></td>
<td>
<a href="../glish/glish.html"
onMouseOver="aMenu.showAipsMenu('glishMenu')">
<font size=2 face=helvetica color="#ffffff">
<b>Glish</b></a></td>
<td>
<a href="../learnmore.html"
onMouseOver="aMenu.showAipsMenu('learnMenu')">
<font size=2 face=helvetica color="#ffffff">
<b>Learn More</b></a></td>
<td>
<a href="../programmer/programmer.html"
onMouseOver="aMenu.showAipsMenu('progMenu')">
<font size=2 face=helvetica color="#ffffff">
<b>Programming</b></a></td>
<td>
<a href="../contactus/contactus.html"
onMouseOver="aMenu.showAipsMenu('commMenu')">
<font size=2 face=helvetica color="#ffffff">
<b>Contact Us</b></a>
</td></tr></table>
<table border=0 width=600>
<tr><td><img src="../gif/aipslogo.gif">
</td><td><font size=2 face=helvetica color="#cc0f0f"><b>VERSIONID</b></font>
</td><td valign=middle>
<font size=2 face=helvetica color="#cc0f0f"><a
href="../newsletters/index.html"><img src="../gif/tri2.gif" border=0> 
 <b>News</a></font></b><P>
<font size=2 face=helvetica color="#cc0f0f"><a
href="../faq/faq.html"><img src="../gif/tri2.gif" border=0>  <b>FAQ</a
></font></b><P>
<font size=2 face=helvetica color="#cc0f0f"><a
href="../search/search.html"><img src="../gif/tri2.gif" border=0>  <b>
Search</a></font></b><P>
<font size=2 face=helvetica color="#cc0f0f"><a href="../aips++.html"><img
src="../gif/tri2.gif" border=0>  <b>Home</a
></font></b><P>

</td></tr></table>
<BR>

<H1>H</H1> 
<P>
<A HREF="#Hanning_smoothing_function">Hanning smoothing function</A><BR>
<A HREF="#hardware_mount">hardware mount</A><BR>
<A HREF="#HasA">HasA</A><BR>
<A HREF="#heliocentric_velocity_system">heliocentric velocity system</A><BR>
<A HREF="#Hermitian_function">Hermitian function</A><BR>
<A HREF="#hexadecimal">hexadecimal</A><BR>
<A HREF="#histogram_modification">histogram modification</A><BR>
<A HREF="#Hogbom_CLEAN">H&ouml;gbom CLEAN</A><BR>
<A HREF="#holography">holography</A><BR>
<A HREF="#homology">homology</A><BR>
<A HREF="#horizon">horizon</A><BR>
<A HREF="#host_computer">host computer</A><BR>
<A HREF="#hour_angle">hour angle</A><BR>
<A HREF="#hour_circle">hour circle</A><BR>
<A HREF="#HPBW">HPBW</A><BR>
<A HREF="#HSV">HSV</A><BR>
<A HREF="#HTML">HTML</A><BR>
<A HREF="#hue">hue</A><BR>
<A HREF="#hybrid_mapping">hybrid mapping</A><BR>
<A HREF="#hydrogen_maser">hydrogen maser</A><BR>
<A HREF="#hypertext">hypertext</A><BR>
</P>
<HR>

<DL>

<DT><STRONG><A NAME="Hanning_smoothing_function">Hanning smoothing function</A></STRONG></DT>

<DD> <P>In the analysis of power spectra, a weight function
<VAR>w</VAR> by which the measured correlation function is multiplied,
in order to reduce that oscillation (<A
HREF="g.html#Gibbs'_phenomenon">Gibbs' phenomenon</A>) in the computed
spectrum which is due to having sampled at only a finite number of
lags.  <VAR>w</VAR>, as a function of lag, is given by </P>

<P><IMG ALIGN=BOTTOM SRC="hanningw.gif"></P>

<P>This is equivalent to convolving the discrete spectrum with the
sequence <VAR>(&#188;, &#189;, &#188;)</VAR>.  </P>

<P>Hanning smoothing sometimes is applied to the <A
HREF="c.html#cross_correlation">cross correlation</A> measurements
obtained in interferometric spectral line observing, to reduce the
effect of sharp bandpass filter cutoffs.  It also is used frequently
in radio astronomical autocorrelation spectroscopy. For more on
smoothing see R. B. Blackman and J. W. Tukey, <CITE>The Measurement of
Power Spectra</CITE>, Dover, New York, 1958.</P>

</DD>

<DT><STRONG><A NAME="hardware_mount">hardware mount</A></STRONG></DT>
<DD>
<P>
The combined acts of installing a computer external storage module,
such as a disk pack or a magnetic tape,
in some electro-mechanical unit
(e.g., a disk drive or a tape drive) that provides computer access
to this data storage medium,
and placing that unit in readiness to be operated under
computer control (e.g., positioning a magnetic tape at
the beginning-of-tape marker).
Compare <A HREF="s.html#software_mount"><EM>software mount</EM></A>.
</P>
</DD>

<DT><STRONG><A NAME="HasA">HasA</A></STRONG></DT> <DD> <P> A term used
when a <A HREF="c.html#class">class</A> has another class embedded
within it (aggregation), or associated with it (a reference or
``pointer'' relationship). Another relationship is <A
HREF="i.html#IsA">IsA</A>.
</P>
</DD>

<DT><STRONG><A NAME="heliocentric_velocity_system">heliocentric velocity system</A></STRONG></DT>
<DD>

<P> Velocity measured relative to the <EM>sun</EM> taken as the focus
of the earth's orbital motion (not relative to the solar system's
barycenter).  Heliocentric and barycentric velocities will differ by
<IMG ALIGN=BOTTOM SRC="leq.gif">0.012 km/s, so the distinction is
sometimes ignored.  </P>

</DD>

<DT><STRONG><A NAME="Hermitian_function">Hermitian function</A></STRONG></DT>
<DD>

<P> A complex-valued function, of one or more real variables, whose
real part is an even function and whose imaginary part is odd.  The
Fourier transform (FT) of a real-valued function is Hermitian, and the
inverse FT of a Hermitian function is real.  </P>

<P>In radio interferometry, each of the radio brightness distributions
<VAR>I(x,y)</VAR>, <VAR>Q(x,y)</VAR>, <VAR>U(x,y)</VAR>, and
<VAR>V(x,y)</VAR> representing 
<A HREF="s.html#Stokes'_parameters">Stokes' parameters</A> is
real-valued.  Their Fourier transforms, the Stokes' <A
HREF="v.html#visibility_function">visibility functions</A>, thus have
the property of <A HREF="c.html#conjugate_symmetry">conjugate
symmetry</A>: </P>

<P><IMG ALIGN=BOTTOM SRC="visymm.gif">,</P>
<P><IMG ALIGN=BOTTOM SRC="vqsymm.gif">,</P>
<P><IMG ALIGN=BOTTOM SRC="vusymm.gif">, and</P>
<P><IMG ALIGN=BOTTOM SRC="vvsymm.gif">.</P>

</DD>

<DT><STRONG><A NAME="hexadecimal">hexadecimal</A></STRONG></DT>
<DD>
<P>
Numeration with a radix (base) of 16.
</P>
</DD>

<DT><STRONG><A NAME="histogram_modification">histogram modification</A></STRONG></DT>
<DD>

<P>The histogram of <A HREF="p.html#pixel">pixel</A> intensities in an
<A HREF="i.html#image">image</A> of a natural scene that has been
linearly <A HREF="q.html#quantization">quantized</A> is usually highly
skewed toward darker levels.  Detail is often imperceptible in the
darker regions when such an image is viewed on a conventional display
with only 8-bit pixels.  Histogram modification can alleviate this
problem by rescaling the original image so that the histogram of pixel
intensities follows some preferred form.  Histogram
<EM>equalization</EM> flattens the output histogram, and may often be
appropriate for astronomical images.  Exponential and hyperbolic
output histograms are also sometimes used, so that a final histogram
"equalization" is performed (approximately) by the quasi-logarithmic
response of the user's retina.  For details, see for example
W.K.Pratt, <CITE>Digital Signal Processing</CITE>, John Wiley and Sons
(1978), p.311-318.</P>

</DD>

<DT><STRONG><A NAME="Hogbom_CLEAN">H&ouml;gbom CLEAN</A></STRONG></DT>
<DD>

<P>A deconvolution algorithm devised by Jan H&ouml;gbom
[<CITE>Aperture synthesis with a non-regular distribution of
interferometer baselines</CITE>, Astron. Astrophys.  Suppl. Ser.,
<STRONG>15</STRONG> (1974) 417-426] for use in radio interferometry.
Denote (the discrete representations of) the <A
HREF="d.html#dirty_image">dirty image</A> by <VAR>g</VAR> and the
dirty beam by <VAR>b</VAR>.  The algorithm iteratively constructs
discrete approximants <VAR>f</VAR><IMG ALIGN=MIDDLE SRC="tiny_n.gif">
to a solution <VAR>f</VAR> of the equation <VAR>b*f = g</VAR> (* here
representing a convolution), starting with an initial approximant that
is zero everywhere.  At the <VAR>n</VAR>'th iteration, one searches
for the peak in the <EM>residual image</EM> <VAR>g - b*f</VAR><IMG
ALIGN=MIDDLE SRC="tiny_n.gif"><IMG ALIGN=MIDDLE SRC="tiny-1.gif">.
</P>

<P>A <IMG ALIGN=BOTTOM SRC="delta.gif">-function component, centered
at the location of the largest residual, and of amplitude <VAR>m</VAR>
(the <em>loop gain</em>) times the largest residual, is added to
<VAR>f</VAR><IMG ALIGN=MIDDLE SRC="tiny_n.gif"><IMG ALIGN=MIDDLE
SRC="tiny-1.gif"> to yield <VAR>f</VAR><IMG ALIGN=MIDDLE
SRC="tiny_n.gif">.  The search over the residual image is restricted
to a region <VAR>A</VAR> termed the <A
HREF="c.html#CLEAN_window">CLEAN window</A>.  The iteration terminates
with an approximate solution <VAR>f</VAR><IMG ALIGN=MIDDLE
SRC="tiny_un.gif"> either when <VAR>N</VAR> reaches some prescribed
iteration limit or when the peak residual (in absolute value) or the
r.m.s.  residual decreases to some prescribed level.  </P>

<P>To diminish any spurious high spatial frequency features in the
solution, <VAR>f</VAR><IMG ALIGN=MIDDLE SRC="tiny_un.gif"> is
convolved with a narrow elliptical Gaussian function <VAR>h</VAR>,
termed the <A HREF="c.html#CLEAN_beam">CLEAN beam</A>.  Generally
<VAR>h</VAR> is chosen by fitting to the central lobe of the 
<A HREF="d.html#dirty_beam">dirty beam</A>. 
Also, one generally adds the
final residual image <VAR>g - b*f</VAR><IMG ALIGN=MIDDLE
SRC="tiny_un.gif"> to the approximate solution <VAR>f</VAR><IMG
ALIGN=MIDDLE SRC="tiny_un.gif">*<VAR>h</VAR>, in order to produce a
final result, termed the <A HREF="c.html#CLEAN_image">CLEAN image</A>,
with a realistic-appearing level of noise.  
</P>

</P>See also 
<A HREF="s.html#super-resolution"><EM>super-resolution</EM></A>.
</P>

</DD>

<DT><STRONG><A NAME="holography">holography</A></STRONG></DT>
<DD>

<P>In radio astronomy, a method for refining the 
<A HREF="f.html#feed">feed</A> and <A HREF="p.html#panel">panel</A>
alignment, thus for improving the 
<A HREF="a.html#aperture_efficiency">aperture efficiency</A> or 
<A HREF="a.html#antenna_gain">antenna gain</A> of a radio telescope.  By
scanning an antenna's beam over a raster around an unresolved radio
source, and using another antenna pointing at the same source as a
reference, information is obtained about the amplitude and phase
distributions of the signal reflected from the antenna surface.
These distributions are used to specify corrections (if
needed) for the focus and alignment of the <A
HREF="f.html#feed">feeds</A> or of the positions of individual panels
in the reflector.  </P>

</DD>

<DT><STRONG><A NAME="homology">homology</A></STRONG></DT>
<DD>

<P>The use of the principle of <EM>homologous deformation</EM> in the
design of <A HREF="a.html#antenna">antenna</A> structures.  Any
massive structure must deflect under its own weight.  As the reflector
of an antenna in a radio telescope must move relative to the direction
of the gravitational force, the deflections must change as the antenna
moves.  An antenna is said to obey the principle of homologous
deformation if the entire structure above the elevation bearings is
designed to maintain the reflector in the shape of a paraboloid of
revolution when pointing to any position between the zenith and the
horizon.  This can be achieved if the focal length of the paraboloid
is allowed to change, and if the electrical axis of the paraboloid is
allowed to move in a slightly different way than the elevation axis
rotates.  (These deviations can be compensated by moving the <A
HREF="f.html#feed">feeds</A>.)  When we achieve the result that the
reflector surface retains the same geometrical shape to a certain
precision at all <A HREF="e.html#elevation">elevation</A> angles,
changing only by a scale factor, the antenna is said to be
<EM>homologous</EM>.  </P>

</DD>

<DT><STRONG><A NAME="horizon">horizon</A></STRONG></DT>
<DD>
<P>The plane perpendicular to the local vertical direction as defined by a
plumb-line.</P>
</DD>

<DT><STRONG><A NAME="host_computer">host computer</A></STRONG></DT>
<DD>

<P>In the parasitic relationship of a computer program or program
package, such as AIPS++, to the computer on which it runs, the latter
is termed the <EM>host computer</EM>.  Also, in the master-slave
relationship of a computer to one of its peripheral devices, such as
an <A HREF="a.html#array_processor">array processor</A> or <A
HREF="c.html#coprocessor">coprocessor</A>, the master may be termed
the <em>host.</em></P>

</DD>

<DT><STRONG><A NAME="hour_angle">hour angle</A></STRONG></DT>
<DD>

<P>The angle <VAR>h</VAR>, measured westward, between the meridian and
the <A HREF="#hour_circle">hour circle</A> passing through a celestial
source.  The hour angle of an object depends both on time and the
observer's location: it can be determined by subtracting the object's
<A HREF="r.html#right_ascension">right ascension</A> from the <A
HREF="l.html#local_sidereal_time">local sidereal time</A>.</P>

</DD>

<DT><STRONG><A NAME="hour_circle">hour circle</A></STRONG></DT>
<DD>

<P>A great circle passing through the celestial poles, i.e. any circle
that is perpendicular to the <A
HREF="c.html#celestial_equator">celestial equator</A>.  </P>

</DD>

<DT><STRONG><A NAME="HPBW">HPBW</A></STRONG></DT>
<DD>

<P>For an antenna beam in any one dimension, the full width in angle
between the directions at which the received or transmitted power
falls to half of its maximum value.  (Synonymous with the <A
HREF="f.html#FWHM">FWHM</A> of the antenna's power pattern in this
dimension.)</P>

</DD>

<DT><STRONG><A NAME="HSV">HSV</A></STRONG></DT>
<DD>

<P>
<A HREF="#hue"><STRONG>H</STRONG>ue</A>, 
<A HREF="s.html#saturation"><STRONG>S</STRONG>aturation</A>, 
<STRONG>V</STRONG>alue: 
A system for describing the physical perception of color, in
terms of tint, perceived narrowness of the spectrum, and 
<A HREF="i.html#intensity">intensity</A> (here substituted for
by "value").  
See also <A HREF="c.html#chromaticity"><EM>chromaticity</EM></A>, 
<A HREF="r.html#RGB"><EM>RGB</EM></A>.
</P>
</DD>

<DT><STRONG><A NAME="HTML">HTML</A></STRONG></DT>
<DD>

<P>The <STRONG>H</STRONG>yper<STRONG>T</STRONG>ext 
<STRONG>M</STRONG>arkup <STRONG>L</STRONG>anguage 
used to format <A HREF="#hypertext">hypertext</A> documents for display 
by <A HREF="b.html#browser">browsers</A> compatible with the 
<A HREF="w.html#World-Wide-Web">World-Wide-Web</A>.  
Built on <A HREF="s.html#SGML">SGML</A>.  
"Tags" consisting of a "&lt;", a "directive", zero or more
parameters and a "&gt;" are embedded in the text. For details on
HTML language standards, manuals, browsers and other tools, consult
Alan Bridle's <A HREF="../html/htm4aips.html"><CITE>HTML Language and Resource Guide
</A></CITE> 
(<KBD>http://aips2.nrao.edu/aips++/docs/html/htm4aips.html</KBD>).
</P>
</DD> 

<DT><STRONG><A NAME="hue">hue</A></STRONG></DT>
<DD>

<P>One of the three basic parameters (<em>hue</em>,
<em>intensity</em>, and <em>saturation</em>) which may be used to
describe the physical perception of the light that reaches one's eye.
Hue, which is also termed <em>tint</em>, or simply <em>color</em>,
refers to the dominant wavelength of the coloration, at a given
location in an image or scene.  The term also may be used to describe
a multimodal color spectrum---e.g., one speaks of a purple hue.
Different spectral distributions of light, of identical intensity and
saturation, are capable of producing identical retinal responses;
these unique responses comprise the set of perceptible hues.  
</P>

<P>Color matching tests have established that there are three basic
types of human retinal receptors, whose peak responses are to red,
green, and blue light.  These are the three <em>primary hues</em> used
in additive color mixing---e.g., in digital image display.  They may
be used to produce all, or virtually all, of the perceptible hues.
</P>

<P>See also 
<A HREF="c.html#chromaticity"><EM>chromaticity</EM></A>, 
<A HREF="r.html#RGB"><EM>RGB</EM></A>.</P></DD> 

<DT><STRONG><A NAME="hybrid_mapping">hybrid mapping</A></STRONG></DT>
<DD>
<P>An <A HREF="a.html#algorithm">algorithm</A> for calibrating radio 
interferometer data, essentially equivalent to the 
<A HREF="s.html#self-calibration">self-calibration</A>
algorithm, except in that it makes explicit use of the 
<A HREF="c.html#closure_phase">closure phase</A> and
<A HREF="c.html#closure_amplitude">closure amplitude</A> relations,
rather than of the relation
</P>
<P><IMG ALIGN=BOTTOM SRC="scgaineq.gif"></P>

<P>relating the observed visibility to the product of the true
visibility and a pair of <A
HREF="a.html#antenna/i.f._gain">antenna/i.f. gains</A>.  Hybrid
mapping, which is used extensively in <A HREF="v.html#VLBI">VLBI</A>
data reduction, is described in A. C. S. Readhead <em>et al.</em>,
<CITE>Mapping radio sources with uncalibrated visibility data</CITE>,
Nature, <STRONG>285</STRONG> (1980) 137-140.  </P>

<P>Either algorithm (assuming that one cares to distinguish them) can
be applied to data obtained with connected-element interferometers
such as the <A HREF="v.html#VLA">VLA</A>, <A
HREF="a.html#ATCA">ATCA</A> or <A HREF="w.html#WSRT">WSRT</A> or with
non-connected-element interferometers such as <A
HREF="m.html#MERLIN">MERLIN</A> and <A HREF="v.html#VLBI">VLBI</A>
arrays.  Any differences in the results produced by the two algorithms
would be attributable primarily to differences in the effective
weighting of the data. (In particular, early implementations of both
algorithms discarded data which could have been used to obtain
overdetermined solutions for the calibration parameters).  </P>
</DD> 

<DT><STRONG><A NAME="hydrogen_maser">hydrogen maser</A></STRONG></DT>
<DD> 

<P>A device that produces a radio signal directly from the transition
between the upper and lower hyperfine levels in the ground state of
atomic hydrogen.  The most stable type of frequency standard or clock
(for time scales of a few seconds to few hours) that is used in <A
HREF="v.html#VLBI">VLBI</A> observations.  For details of the hydrogen
maser that is used as a Master Clock by the U.S. Naval Observatory,
see <A
HREF="http://tycho.usno.navy.mil/maser.html"><kbd>http://tycho.usno.navy.mil/maser.html</kbd></A>.
</P>

</DD>

<DT><STRONG><A NAME="hypertext">hypertext</A></STRONG></DT>
<DD>

<P>A cluster of documents (or "nodes") containing cross-references or
"links" which let a reader move easily from one document to another
using an interactive <A HREF="b.html#browser">browser</A>.</P>

</DD>

</DL>
<HR>

<P>You could now go back to the:</P>
<UL>
<LI>previous document (use <KBD>back</KBD> on your browser)</LI>
<LI><A HREF="glossary.html">Glossary Index</A></LI>
</UL>
<HR>

<P>Copyright &#169; 1995,1996,1999,2000 Associated Universities Inc., Washington, D.C.</P>
<ADDRESS>abridle@nrao.edu, 26 August 1996, 17:30 EDT</ADDRESS>
</BODY>
</HTML>
