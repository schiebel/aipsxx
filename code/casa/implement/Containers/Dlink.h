//# Dlink.h: Doubly linked list primitive
//# Copyright (C) 1993,1994,1995,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: Dlink.h,v 19.8 2005/06/18 21:19:14 ddebonis Exp $

#ifndef CASA_DLINK_H
#define CASA_DLINK_H

#include <casa/aips.h>
#include <casa/Containers/Link.h>

namespace casa { //# NAMESPACE CASA - BEGIN

// <summary><b>Deprecated</b> use <linkto class=Link>Link</linkto> instead</summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
// <h2>Deprecated use <linkto class=Link><src>Link</src></linkto> instead.</h2>
//
template<class t> class Dlink : public Link<t> {
  public:
    Dlink(t e,Dlink<t> *p=0,Dlink<t> *n=0);
    Dlink<t> *&prev() {return (Dlink<t>*&) (this->Prev);}
    const Dlink<t> *prev() const {return (Dlink<t>*) (this->Prev);}
};


} //# NAMESPACE CASA - END

#ifndef AIPS_NO_TEMPLATE_SRC
#include <casa/Containers/Dlink.cc>
#endif //# AIPS_NO_TEMPLATE_SRC
#endif
