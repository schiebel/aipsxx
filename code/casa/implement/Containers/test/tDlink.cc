//# tDlink.cc: This program tests the doubly linked list class
//# Copyright (C) 1993,1994,1995,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: tDlink.cc,v 19.3 2004/11/30 17:50:16 ddebonis Exp $

//# Includes

#include <casa/Containers/Dlink.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
main() {
  Dlink<int> *hed = new Dlink<int>(23);

  hed = new Dlink<int>(12,0,hed);
  hed = new Dlink<int>(19,0,hed);
  hed = new Dlink<int>(82,0,hed);
  hed = new Dlink<int>(71,0,hed);
  hed = new Dlink<int>(3,0,hed);
  hed = new Dlink<int>(10,0,hed);

  Dlink<int> *cur = hed;
  int forward = 1;
  cout << ">";
  while (cur) {
    cout << (*cur).val() << ".";
    if (forward)
      if ((*cur).next())
	cur = (Dlink<int>*) (*cur).next();
      else {
	cout << endl << "<";
	forward = 0;
      }
    else
      cur = (*cur).prev();
  }
  cout << endl;

  cur = (Dlink<int>*) hed->next()->next()->next()->next();
  Dlink<int> *prev = cur->prev();

  cur = new Dlink<int>(103,prev,cur);
  cur = new Dlink<int>(111,prev,cur);
  cur = new Dlink<int>(172,prev,cur);

  cur = hed;
  forward = 1;
  Dlink<int> *end = 0;
  cout << ">";
  while (cur) {
    cout << (*cur).val() << ".";
    if (forward)
      if ((*cur).next())
	cur = (Dlink<int>*) (*cur).next();
      else {
	cout << endl << "<";
	end = cur;
	forward = 0;
      }
    else
      cur = (*cur).prev();
  }
  cout << endl;

  end = new Dlink<int>(203,end);
  end = new Dlink<int>(221,end);
  end = new Dlink<int>(201,end);

  cur = hed;
  forward = 1;
  cout << ">";
  while (cur) {
    cout << (*cur).val() << ".";
    if (forward)
      if ((*cur).next())
	cur = (Dlink<int>*) (*cur).next();
      else {
	cout << endl << "<";
	forward = 0;
      }
    else
      cur = (*cur).prev();
  }
  cout << endl;

  cur = hed;
  while (cur) {
    (*cur).val()++;
    cur = (Dlink<int>*) (*cur).next();
  }

  cur = hed;
  forward = 1;
  cout << ">";
  while (cur) {
    cout << (*cur).val() << ".";
    if (forward)
      if ((*cur).next())
	cur = (Dlink<int>*) (*cur).next();
      else {
	cout << endl << "<";
	forward = 0;
      }
    else
      cur = (*cur).prev();
  }
  cout << endl;

  return 0;
}
