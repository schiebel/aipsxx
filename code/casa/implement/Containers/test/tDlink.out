>10.3.71.82.19.12.23.
<23.12.19.82.71.3.10.
>10.3.71.82.172.111.103.19.12.23.
<23.12.19.103.111.172.82.71.3.10.
>10.3.71.82.172.111.103.19.12.23.203.221.201.
<201.221.203.23.12.19.103.111.172.82.71.3.10.
>11.4.72.83.173.112.104.20.13.24.204.222.202.
<202.222.204.24.13.20.104.112.173.83.72.4.11.
